app/src/org/commcare/android/database/app/models/UserKeyRecord.java

```diff
@@ -208,6 +267,56 @@ public boolean isPasswordValid(String password) {
                 hash.equals(UserKeyRecord.generatePwdHash(password, UserKeyRecord.extractSalt(hash))));
     }
 
+    public void assignPinToRecord(String pin, String password) {
+        this.passwordWrappedByPin = wrapPasswordWithPin(password, pin);
+    }
+
+    public byte[] getWrappedPassword() {
+        return passwordWrappedByPin;
+    }
+
+    public boolean hasPinSet() {
+        return passwordWrappedByPin.length > 0;
+    }
+
+    public boolean isPinValid(String pin) {
+        // Unwrap wrapped password with the PIN, and then check if the resulting password is correct
+        return isPasswordValid(getUnhashedPasswordViaPin(pin));
+    }
+
+    /**
+     * Returns the un-hashed password that was wrapped by the given PIN, or null if the given PIN
+     * is not valid to unwrap the wrapped password
+     */
+    public String getUnhashedPasswordViaPin(String pin) {
+        return unwrapPasswordWithPin(this.passwordWrappedByPin, pin);
+    }
+
+    /**
+     * Must be called with 1 of the 2 values set to null, which indicates that we should check
+     * based upon the other
+     */
+    public boolean isPasswordOrPinValid(String password, String pin) {
+        if (pin != null) {
+            return isPinValid(pin);
+        } else {
+            return isPasswordValid(password);
+        }
+    }
+
+    public static byte[] wrapPasswordWithPin(String password, String pin) {
+        return CryptUtil.wrapByteArrayWithString(password.getBytes(), pin);
+    }
+
+    public static String unwrapPasswordWithPin(byte[] wrappedPassword, String pin) {
```

Comment:
can be `private`. Consider moving into code to `getUnhashedPasswordViaPin` since that is the only place it is called and the only thing that `getUnhashedPasswordViaPin` does is call this helper.